From 9c6716ed75fe96c1b3ea2b9a7e26cde1e9994973 Mon Sep 17 00:00:00 2001
From: Pietro Delugas <pdelugas@sissa.it>
Date: Mon, 26 Feb 2018 09:43:52 +0100
Subject: [PATCH] changed the initialization of IO error codes.

Initializing the IO error codes each time that a xml file
in open for reading may become cumbersome in parallel application
with many processes opening scratch file for tests. In this case
it is preferable to run setup_io just with one node and propagate the
results to all other nodes.
I have redefined setup_io as an interface which refers to 2 module
procedures. setup_io_scracth which is the former  setup_io routine and
a setup_io_with_data where err_code eor_code and eof_code are provided
as arguments with which io_err, io_eor and io_eof are initialized.
I have added a private logical variable io_init that is toggled to true by
setup_io on exit. setup_io checks the io_init variable when starting and
in true returns without no further actions.
---
 common/m_common_io.F90 | 23 +++++++++++++++++++++--
 1 file changed, 21 insertions(+), 2 deletions(-)

diff --git a/common/m_common_io.F90 b/common/m_common_io.F90
index a15a25db..e7724521 100644
--- a/common/m_common_io.F90
+++ b/common/m_common_io.F90
@@ -11,6 +11,11 @@ module m_common_io
   integer, save :: io_eor
   integer, save :: io_eof
   integer, save :: io_err
+  logical, save :: io_init=.false.
+
+  interface  setup_io
+      module procedure setup_io_scratch, setup_io_with_data
+  end interface setup_io
 
   public :: io_eor
   public :: io_eof
@@ -20,9 +25,23 @@ module m_common_io
 
 contains
 
-  subroutine setup_io()
+  subroutine setup_io_scratch()
+    if (io_init) return 
     call find_eor_eof(io_eor, io_eof)
-  end subroutine setup_io
+    io_init = .true. 
+  end subroutine setup_io_scratch
+
+  subroutine setup_io_with_data(err_code, eor_code, eof_code)
+     implicit none
+     integer,intent(in)  :: err_code, eor_code, eof_code
+     if (io_init) return 
+     io_err = err_code
+     io_eor  = eor_code
+     io_eof  = eof_code
+     io_init = .true.
+   end subroutine setup_io_with_data 
+
+  
 
 
   subroutine get_unit(lun,iostat)
